optimade/server/config.py

```diff
@@ -36,14 +35,74 @@ class LogLevel(Enum):
     CRITICAL = "critical"
 
 
+def config_file_settings(settings: BaseSettings) -> Dict[str, Any]:
+    """Configuration file settings source.
+
+    Based on the example in the
+    [pydantic documentation](https://pydantic-docs.helpmanual.io/usage/settings/#adding-sources),
+    this function loads ServerConfig settings from a configuration file.
+
+    The file must be of either type JSON or YML/YAML.
+
+    Parameters:
+        settings: The `pydantic.BaseSettings` class using this function as a
+            `pydantic.SettingsSourceCallable`.
+
+    Returns:
+        Dictionary of settings as read from a file.
+
+    """
+    import json
+    import os
+    import yaml
+
+    encoding = settings.__config__.env_file_encoding
+    config_file = Path(os.getenv("OPTIMADE_CONFIG_FILE", DEFAULT_CONFIG_FILE_PATH))
+
+    res = {}
+    if config_file.is_file():
+        config_file_content = config_file.read_text(encoding=encoding)
+
+        try:
+            res = json.loads(config_file_content)
+        except json.JSONDecodeError as json_exc:
+            try:
+                # This can essentially also load JSON files, as JSON is a subset of YAML v1,
+                # but I suspect it is not as rigorous
+                res = yaml.safe_load(config_file_content)
+            except yaml.YAMLError as yaml_exc:
+                warnings.warn(
+                    f"Unable to parse config file {config_file} as JSON and as YAML, using the "
+                    "default settings instead..\n"
+                    f"Errors:\n  JSON:\n{json_exc}.\n\n  YAML:\n{yaml_exc}"
+                )
+    else:
+        warnings.warn(
+            f"Unable to find config file at {config_file}, using the default settings instead."
+        )
+
+    if res:
+        # Set "config_file" parameter to acknowledge a config file has been used.
+        res["config_file"] = str(config_file)
+    elif res is None:
+        # This can happen if the yaml loading doesn't succeed properly.
+        res = {}
+
+    return res
+
+
 class ServerConfig(BaseSettings):
     """This class stores server config parameters in a way that
     can be easily extended for new config file types.
 
     """
 
     config_file: Optional[str] = Field(
-        None, description="File to load alternative defaults from"
+        None,
+        description=(
+            "File to load alternative defaults from, (default location to look: "
+            f"{DEFAULT_CONFIG_FILE_PATH})"
```

Comment:
```suggestion
            "File to load alternative defaults from, (default location to look: "
            f"{DEFAULT_CONFIG_FILE_PATH}, controllable by the OPTIMADE_CONFIG_FILE environment variable.)"
```

Reply:
Should we even keep this setting any more? We could just handle `OPTIMADE_CONFIG_FILE` env var individually

Reply:
Indeed. But I was thinking for testing purposes and more it might be nice to have the information of where the config file is located and such directly from the `CONFIG`. _But_ there is really no reason to keep it, honestly. It's just redundant information.
You want to remove it?

Reply:
Ah. And it's nice to use for documentational reasons as well, since we cannot display `DEFAULT_CONFIG_FILE_PATH` in the API documentation overview :/

Reply:
I've removed it in the latest commit e8641f2. In doing so I've also added some documentation for the `DEFAULT_CONFIG_FILE_PATH` variable hardcoding in its value. I think that's fine at this point. One can easily see and update it if necessary in the future.

I've also added a bit more documentation sugar to the `Enum` classes.